docs/source/torch.compiler_cudagraph_trees.rst

```diff
@@ -112,8 +116,165 @@ Let’s say we are benchmarking running inference with the following code:
+  their contents because the recorded memory address will be the same as the executed memory address.
+* **Tensors which are prior outputs from CUDAGraph Trees**: Because the output tensor addresses of a cudagraph are fixed, if we run CUDAGraph1,
+  then run CUDAGraph2, the inputs which came from CUDAGraph1 into CUDAGraph2 will have a fixed memory address. These inputs, like parameters and
+  buffers, do not require copying over to a static address tensor. We check to make sure that these inputs are stable at runtime, and if they're
+  not we will re-record.
+
+CUDAGraph Trees support input mutation on parameters and buffers, and tensors which are prior outputs from CUDAGraph Trees. For mutation on inputs
+from eager, CUDAGraph Trees will run the function without CUDAGraph and emit *skipping due to mutated inputs* log. The following example shows
+CUDAGraph Trees' support for tensors which are prior outputs from CUDAGraph Trees.
+
+.. code-block:: python
+
+    import torch
+
+    @torch.compile(mode="reduce-overhead")
+    def foo(x):
+        return x + 1
+
+    @torch.compile(mode="reduce-overhead")
+    def mut(x):
+        return x.add_(2)
+
+    for i in range(3):
+        torch.compiler.cudagraph_mark_step_begin()
+        inp = torch.rand([4], device="cuda")
+
+        # CUDAGraph is applied since `foo` does not mutate `inp`
+        tmp = foo(inp)
+        # Although `mut` mutates `tmp`, which is an output of a CUDAGraph
+        # managed function. So CUDAGraph is still applied.
+        mut(tmp)
+
+
+    torch.compiler.cudagraph_mark_step_begin()
+    inp = torch.rand([4], device="cuda")
+
+    tmp = foo(inp)
+    # While `tmp` is a CUDAGraph Tree managed function's output, `tmp.clone()`
+    # is not. So CUDAGraph is not applied to `mut` and there is a log
+    # `skipping cudagraphs due to mutated inputs`
+    mut(tmp.clone())
+
+To enable CUDAGraph Trees for a function mutating inputs from eager, please re-write
+the function to avoid input mutation, as illustrated below:
+
+
+.. code-block:: python
+
+    def foo(x, y):
+        # return x + 1 as an additional output tensor
+        tmp = x + 1
+        return tmp + y, tmp
+
+
+Dynamic Shape Support
+---------------------
+
+`Dynamic shape <https://pytorch.org/docs/stable/torch.compiler_dynamic_shapes.html>`_ means that an input tensor has different shapes across
+function calls. Since CUDAGraph requires fixed tensor addresses, CUDAGraph Trees re-record CUDAGraph for every unique shape of an input tensor.
+This leads to multiple CUDAGraphs for a single inductor graph. When there are limited shapes (e.g., batch sizes in inference), it is profitable
+to re-record CUDAGraphs. However, if input tensor shapes change on every invocation, re-recording CUDAGraph may not be profitable. In this case,
+we suggest to pad input tensors to a few fixed shapes to still enjoy benefits from CUDAGraph.
```

Comment:
Add reference to cudagraph_skip_dynamic_graphs ? if they wish to cudagraph static subgraphs but skip symbolic ones ? I would also add a mention that Nvidia uses 64kb of device memory per kernel launch in cudagraphs, up until cuda 12.4 driver 550+. With many re-recordings this can be significant.